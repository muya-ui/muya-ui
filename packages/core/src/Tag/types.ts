import { ReactNode } from 'react';

import { ICustomStyleBaseProps, ISizeSpecBaseProps } from '../types';

export type ITagShape = 'circle' | 'rect';
/**
 * Tag style keys
 * @styles childrenWrapper 子节点容器
 * @styles closeIcon 关闭按钮
 */
export type ITagStyleKeys = 'childrenWrapper' | 'closeIcon';

interface ITagBaseProps {
  /**
   * 最大宽度
   */
  maxWidth?: number;
  /**
   * 形状
   * @default 'circle'
   */
  shape?: ITagShape;
  /**
   * 禁用状态
   * @default false
   */
  disabled?: boolean;
  /**
   * 是否有边框
   * @default false
   */
  bordered?: boolean;
  /**
   * 边框颜色
   */
  borderColor?: string;
  /**
   * 子节点
   */
  children: ReactNode;
}

export interface ITagPureProps
  extends ISizeSpecBaseProps,
    ITagBaseProps,
    ICustomStyleBaseProps<ITagStyleKeys> {
  /** 颜色 */
  color?: string;
  /**
   * 彩色低显示亮色
   * @default true
   */
  colorInverse?: boolean;
  /**
   * 是否可关闭
   * @default false
   **/
  closable?: boolean;
  /**
   * hover 是否高亮标签
   * @default true
   **/
  hoverable?: boolean;
  /**
   * 是否可见
   * @default true
   **/
  visible?: boolean;
  /**
   * 关闭的回调
   */
  onClose?: React.MouseEventHandler;
}

export type ITagProps = ITagPureProps & React.HTMLAttributes<HTMLSpanElement>;

export interface ICheckableTagPureProps extends ISizeSpecBaseProps, ITagBaseProps {
  /**
   * 是否选中
   * @default false
   */
  checked: boolean;
  /**
   * 改变的回调
   */
  onChange?: (checked: boolean) => void;
  /**
   * 子节点
   */
  children: ReactNode;
}

export type ICheckableTagProps = ICheckableTagPureProps & React.HTMLAttributes<HTMLSpanElement>;
