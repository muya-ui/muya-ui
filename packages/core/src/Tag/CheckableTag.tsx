import React, { Ref } from 'react';
import styled from 'styled-components';

import { useEventCallback } from '@muya-ui/utils';

import memoForwardRef from '../utils/memoForwardRef';
import { withThemeForStyled } from '../utils/withTheme';
import { checkableTagCss, tagCss } from './styled';
import { ICheckableTagProps } from './types';

const BaseNode = styled.span``;

const PureCheckableTag = memoForwardRef((props: ICheckableTagProps, ref: Ref<HTMLSpanElement>) => {
  const {
    shape,
    size = 'm',
    bordered,
    disabled,
    borderColor,
    maxWidth,
    checked,
    onChange,
    onClick,
    ...restProps
  } = props;
  const handleClick = useEventCallback(
    (e: React.MouseEvent<HTMLSpanElement>) => {
      if (!disabled) {
        /* istanbul ignore else */
        if (onClick) {
          onClick(e);
        }
        /* istanbul ignore else */
        if (onChange) {
          onChange(!checked);
        }
      } else {
        e.preventDefault();
      }
    },
    [checked, onChange],
  );
  return <BaseNode ref={ref} {...restProps} onClick={handleClick}></BaseNode>;
});

const CheckableTag = styled(PureCheckableTag)`
  ${tagCss};
  ${checkableTagCss}
`;

export default withThemeForStyled(CheckableTag);
