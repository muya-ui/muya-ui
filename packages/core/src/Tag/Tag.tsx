import React, { Ref, useCallback, useRef, useState, useMemo } from 'react';
import styled from 'styled-components';

import { CloseIcon as DefaultCloseIcon } from '@muya-ui/theme-light';

import Animation from '../Animation';
import memoForwardRef from '../utils/memoForwardRef';
import useAdjustMiniFont from '../utils/useAdjustMiniFont';
import useStyles from '../utils/useStyles';
import useTheme from '../utils/useTheme';
import { withThemeForStyled } from '../utils/withTheme';
import { StyledCloseIcon, StyledTagText, tagCss } from './styled';
import { ITagProps } from './types';

const BaseNode = styled.span``;
const defaultStyles = {
  childrenWrapper: '',
  closeIcon: '',
};

const PureTag = memoForwardRef((props: ITagProps, ref: Ref<HTMLSpanElement>) => {
  const {
    visible = true,
    closable,
    children,
    color,
    shape,
    size = 'm',
    bordered,
    colorInverse,
    disabled,
    borderColor,
    maxWidth,
    styles,
    onClick,
    onClose,
    ...restProps
  } = props;
  const theme = useTheme();
  const {
    components: { Tag: token },
  } = theme;
  const isVisibleControlled = 'visible' in props;
  const fontSize = token.fontSize[size!];
  const textRef = useRef<HTMLSpanElement>(null);
  const { closeIcon: CloseIcon = DefaultCloseIcon } = theme.components.Tag;
  const [visibleState, setVisibleState] = useState(visible);
  useAdjustMiniFont(textRef, fontSize);

  const innerStyles = useStyles('tag', defaultStyles, styles);

  const handleClick = useCallback(
    (e: React.MouseEvent<HTMLSpanElement>) => {
      if (!disabled) {
        if (onClick) {
          onClick(e);
        }
      } else {
        e.preventDefault();
      }
    },
    [disabled, onClick],
  );

  const handleCloseIconClick = useCallback(
    (e: React.MouseEvent<SVGSVGElement>) => {
      e.stopPropagation();
      if (!disabled) {
        /* istanbul ignore else */
        if (onClose) {
          onClose(e);
        }
        if (e.defaultPrevented) {
          return;
        }
        /* istanbul ignore else */
        if (!isVisibleControlled) {
          setVisibleState(false);
        }
      } else {
        e.preventDefault();
      }
    },
    [disabled, isVisibleControlled, onClose],
  );

  const handleEntered = useCallback((node: HTMLElement) => {
    /* istanbul ignore next */
    if (node && node.style) {
      node.style.display = '';
    }
  }, []);

  const handleExited = useCallback((node: HTMLElement) => {
    /* istanbul ignore next */
    if (node && node.style) {
      node.style.display = 'none';
    }
  }, []);

  if (isVisibleControlled && visible !== visibleState) {
    setVisibleState(visible!);
  }

  const TagNode = useMemo(
    () => (
      <BaseNode ref={ref} onClick={handleClick} {...restProps}>
        <StyledTagText ref={textRef} {...innerStyles.childrenWrapper}>
          {children}
        </StyledTagText>
        {closable && (
          <StyledCloseIcon
            {...innerStyles.closeIcon}
            theme={theme}
            $size={size}
            $color={color}
            $colorInverse={colorInverse}
          >
            <CloseIcon onClick={handleCloseIconClick} />
          </StyledCloseIcon>
        )}
      </BaseNode>
    ),
    [
      children,
      closable,
      color,
      colorInverse,
      handleClick,
      handleCloseIconClick,
      innerStyles.childrenWrapper,
      innerStyles.closeIcon,
      ref,
      restProps,
      size,
      theme,
    ],
  );

  if (disabled) {
    return TagNode;
  }

  return (
    <Animation.Fade in={visibleState} onEntered={handleEntered} onExited={handleExited}>
      {TagNode}
    </Animation.Fade>
  );
});

const Tag = styled(PureTag)`
  ${tagCss};
`;

export default withThemeForStyled(Tag);
