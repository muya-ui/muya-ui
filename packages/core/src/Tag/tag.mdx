---
name: Tag 标签
menu: 组件
route: /components/tag
autoDemos: true
parent: 数据展示
---

import { DocTagStyle } from './demos//styled';

<DocTagStyle />

# Tag 标签

进行标记和分类的小标签。

## 何时使用

- 用于标记事物的属性和维度。
- 进行分类。

## 组件

- `Tag`：标签。
- `CheckableTag`：可选择的标签。

{Demos}

### API

### ITagProps

<Props propsName="ITagProps" />

### ICheckableTagProps

<Props propsName="ICheckableTagProps" />

### Styles

#### ITagStyleKeys

<Styles typeName="ITagStyleKeys" componentName="tag" />
