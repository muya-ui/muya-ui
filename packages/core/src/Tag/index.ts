import CheckableTag from './CheckableTag';
import Tag from './Tag';

export * from './types';
export default Tag;
export { Tag, CheckableTag };
