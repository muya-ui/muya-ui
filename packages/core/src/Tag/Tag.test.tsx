import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import CheckableTag from './CheckableTag';
import Tag from './Tag';

test('测试普通渲染', () => {
  const tree = renderer
    .create(
      <>
        <Tag>Test</Tag>
        <Tag closable>Test</Tag>
        <Tag color="#666" closable>
          Test
        </Tag>
        <Tag color="#666">Test</Tag>
        <Tag color="#666" bordered>
          Test
        </Tag>
        <Tag color="#666" disabled>
          Test
        </Tag>
        <Tag color="#666" colorInverse closable>
          Test
        </Tag>
        <Tag visible={false} maxWidth={100}>
          Test
        </Tag>
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
  const checkableTree = renderer
    .create(
      <>
        <CheckableTag checked={false}>Test</CheckableTag>
        <CheckableTag checked>Test</CheckableTag>
      </>,
    )
    .toJSON();
  expect(checkableTree).toMatchSnapshot();
});

test('测试 Tag visible 受控逻辑', () => {
  const wrapper = mount(<Tag visible={false}>Test</Tag>);
  wrapper.setProps({ visible: true });
});

test('测试 Tag closeIconClick 逻辑', () => {
  const onClose = jest.fn();
  const wrapper = mount(
    <Tag closable onClose={onClose}>
      Test
    </Tag>,
  );
  wrapper.find('svg').simulate('click');
  expect(onClose).toBeCalledTimes(1);
  wrapper.find('svg').simulate('click', { defaultPrevented: true });
  expect(onClose).toBeCalledTimes(2);
});

test('测试 CheckableTag onChange 逻辑', () => {
  const onChange = jest.fn();
  const onClick = jest.fn();
  const wrapper = mount(
    <CheckableTag checked onChange={onChange} onClick={onClick}>
      Test
    </CheckableTag>,
  );
  wrapper.simulate('click');
  wrapper.simulate('mousedown');
  wrapper.simulate('mouseup');
  expect(onChange).toBeCalledTimes(1);
});
