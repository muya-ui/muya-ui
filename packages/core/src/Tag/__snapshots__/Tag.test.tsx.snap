// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试普通渲染 1`] = `
Array [
  .c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: none;
  color: rgba(102,102,102,1);
  background-color: rgba(248,248,248,1);
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  padding: 0 10px;
  margin-right: 8px;
  font-size: 12px;
  line-height: 1;
  height: 24px;
  white-space: nowrap;
  border-radius: 999rem;
  cursor: default;
  vertical-align: middle;
}

.c0 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0:hover {
  color: rgba(26,122,248,1);
  background-color: rgba(26,122,248,0.12);
  opacity: 1;
}

.c0 a:hover {
  color: rgba(26,122,248,1);
}

.c2 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c3 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c4 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c5 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c6 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c7 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<span
    className="c0"
    onClick={[Function]}
    style={
      Object {
        "opacity": 1,
        "transitionDuration": "300ms",
        "transitionProperty": "opacity",
        "transitionTimingFunction": "cubic-bezier(0.0, 0, 0.2, 1)",
      }
    }
  >
    <span
      className="c1 muya-tag-children-wrapper"
    >
      Test
    </span>
  </span>,
  .c3 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c2 {
  margin-left: 6px;
  cursor: pointer;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 10px;
  height: 10px;
  color: rgba(204,204,204,1);
}

.c2:hover {
  color: rgba(191,191,191,1);
}

.c2:active {
  color: rgba(179,179,179,1);
}

.c2 svg {
  width: 10px;
  height: 10px;
}

.c4 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: none;
  color: rgba(102,102,102,1);
  background-color: rgba(248,248,248,1);
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  padding: 0 10px;
  margin-right: 8px;
  font-size: 12px;
  line-height: 1;
  height: 24px;
  white-space: nowrap;
  border-radius: 999rem;
  cursor: default;
  vertical-align: middle;
}

.c0 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c5 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c6 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c7 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c8 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c9 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<span
    className="c0"
    onClick={[Function]}
    style={
      Object {
        "opacity": 1,
        "transitionDuration": "300ms",
        "transitionProperty": "opacity",
        "transitionTimingFunction": "cubic-bezier(0.0, 0, 0.2, 1)",
      }
    }
  >
    <span
      className="c1 muya-tag-children-wrapper"
    >
      Test
    </span>
    <span
      className="c2 muya-tag-close-icon"
    >
      <svg
        aria-hidden="false"
        className="c3"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 1024 1024"
      >
        <path
          d="M512 421.653L871.04 62.592l90.517 90.507-359.04 359.04 358.87 358.901-90.507 90.517L512 602.667 154.507 960.16 64 869.643 421.493 512.16 63.84 154.507 154.357 64 512 421.653z"
        />
      </svg>
    </span>
  </span>,
  .c3 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c2 {
  margin-left: 6px;
  cursor: pointer;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 10px;
  height: 10px;
  color: rgba(204,204,204,1);
}

.c2:hover {
  color: rgba(191,191,191,1);
}

.c2:active {
  color: rgba(179,179,179,1);
}

.c2 svg {
  width: 10px;
  height: 10px;
}

.c4 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c5 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: none;
  color: rgba(255,255,255,1);
  background-color: #666;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  padding: 0 10px;
  margin-right: 8px;
  font-size: 12px;
  line-height: 1;
  height: 24px;
  white-space: nowrap;
  border-radius: 999rem;
  cursor: default;
  vertical-align: middle;
}

.c0 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c6 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c7 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c8 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c9 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<span
    className="c0"
    onClick={[Function]}
    style={
      Object {
        "opacity": 1,
        "transitionDuration": "300ms",
        "transitionProperty": "opacity",
        "transitionTimingFunction": "cubic-bezier(0.0, 0, 0.2, 1)",
      }
    }
  >
    <span
      className="c1 muya-tag-children-wrapper"
    >
      Test
    </span>
    <span
      className="c2 muya-tag-close-icon"
    >
      <svg
        aria-hidden="false"
        className="c3"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 1024 1024"
      >
        <path
          d="M512 421.653L871.04 62.592l90.517 90.507-359.04 359.04 358.87 358.901-90.507 90.517L512 602.667 154.507 960.16 64 869.643 421.493 512.16 63.84 154.507 154.357 64 512 421.653z"
        />
      </svg>
    </span>
  </span>,
  .c2 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c3 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c4 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: none;
  color: rgba(255,255,255,1);
  background-color: #666;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  padding: 0 10px;
  margin-right: 8px;
  font-size: 12px;
  line-height: 1;
  height: 24px;
  white-space: nowrap;
  border-radius: 999rem;
  cursor: default;
  vertical-align: middle;
}

.c0 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0:hover {
  color: rgba(255,255,255,1);
  background-color: #666;
  opacity: 0.9;
}

.c0 a:hover {
  color: rgba(26,122,248,1);
}

.c5 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c6 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c7 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<span
    className="c0"
    onClick={[Function]}
    style={
      Object {
        "opacity": 1,
        "transitionDuration": "300ms",
        "transitionProperty": "opacity",
        "transitionTimingFunction": "cubic-bezier(0.0, 0, 0.2, 1)",
      }
    }
  >
    <span
      className="c1 muya-tag-children-wrapper"
    >
      Test
    </span>
  </span>,
  .c2 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c3 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c4 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c5 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: none;
  color: rgba(255,255,255,1);
  background-color: #666;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  padding: 0 10px;
  margin-right: 8px;
  font-size: 12px;
  line-height: 1;
  height: 24px;
  white-space: nowrap;
  border-radius: 999rem;
  cursor: default;
  vertical-align: middle;
  border: 1px solid rgba(238,238,238,1);
}

.c0 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0:hover {
  color: rgba(255,255,255,1);
  background-color: #666;
  opacity: 0.9;
}

.c0 a:hover {
  color: rgba(26,122,248,1);
}

.c6 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c7 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<span
    className="c0"
    onClick={[Function]}
    style={
      Object {
        "opacity": 1,
        "transitionDuration": "300ms",
        "transitionProperty": "opacity",
        "transitionTimingFunction": "cubic-bezier(0.0, 0, 0.2, 1)",
      }
    }
  >
    <span
      className="c1 muya-tag-children-wrapper"
    >
      Test
    </span>
  </span>,
  .c2 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c3 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c4 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c5 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c6 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: none;
  color: rgba(255,255,255,1);
  background-color: #666;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  padding: 0 10px;
  margin-right: 8px;
  font-size: 12px;
  line-height: 1;
  height: 24px;
  white-space: nowrap;
  border-radius: 999rem;
  cursor: default;
  vertical-align: middle;
  opacity: 0.3;
  cursor: default;
}

.c0 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c7 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<span
    className="c0"
    onClick={[Function]}
  >
    <span
      className="c1 muya-tag-children-wrapper"
    >
      Test
    </span>
  </span>,
  .c3 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c2 {
  margin-left: 6px;
  cursor: pointer;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 10px;
  height: 10px;
  color: rgba(255,255,255,1);
}

.c2 svg {
  width: 10px;
  height: 10px;
}

.c4 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c5 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: none;
  color: rgba(255,255,255,1);
  background-color: #666;
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  padding: 0 10px;
  margin-right: 8px;
  font-size: 12px;
  line-height: 1;
  height: 24px;
  white-space: nowrap;
  border-radius: 999rem;
  cursor: default;
  vertical-align: middle;
}

.c0 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c6 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c7 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c8 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c9 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<span
    className="c0"
    onClick={[Function]}
    style={
      Object {
        "opacity": 1,
        "transitionDuration": "300ms",
        "transitionProperty": "opacity",
        "transitionTimingFunction": "cubic-bezier(0.0, 0, 0.2, 1)",
      }
    }
  >
    <span
      className="c1 muya-tag-children-wrapper"
    >
      Test
    </span>
    <span
      className="c2 muya-tag-close-icon"
    >
      <svg
        aria-hidden="false"
        className="c3"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 1024 1024"
      >
        <path
          d="M512 421.653L871.04 62.592l90.517 90.507-359.04 359.04 358.87 358.901-90.507 90.517L512 602.667 154.507 960.16 64 869.643 421.493 512.16 63.84 154.507 154.357 64 512 421.653z"
        />
      </svg>
    </span>
  </span>,
  .c2 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c3 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c4 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c5 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c6 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c7 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 100px;
  color: rgba(102,102,102,1);
  background-color: rgba(248,248,248,1);
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  padding: 0 10px;
  margin-right: 8px;
  font-size: 12px;
  line-height: 1;
  height: 24px;
  white-space: nowrap;
  border-radius: 999rem;
  cursor: default;
  vertical-align: middle;
}

.c0 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0:hover {
  color: rgba(26,122,248,1);
  background-color: rgba(26,122,248,0.12);
  opacity: 1;
}

.c0 a:hover {
  color: rgba(26,122,248,1);
}

<span
    className="c0"
    onClick={[Function]}
    style={
      Object {
        "opacity": 0,
        "transitionDuration": "300ms",
        "transitionProperty": "opacity",
        "transitionTimingFunction": "cubic-bezier(0.4, 0, 0.2, 1)",
        "visibility": "hidden",
      }
    }
  >
    <span
      className="c1 muya-tag-children-wrapper"
    >
      Test
    </span>
  </span>,
]
`;

exports[`测试普通渲染 2`] = `
Array [
  .c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: none;
  color: rgba(102,102,102,1);
  background-color: rgba(248,248,248,1);
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  padding: 0 10px;
  margin-right: 8px;
  font-size: 12px;
  line-height: 1;
  height: 24px;
  white-space: nowrap;
  border-radius: 999rem;
  cursor: default;
  vertical-align: middle;
  color: rgba(102,102,102,1);
  background-color: rgba(248,248,248,1);
}

.c0 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0:hover {
  color: rgba(26,122,248,1);
  background-color: rgba(26,122,248,0.12);
  opacity: 1;
}

.c0 a:hover {
  color: rgba(26,122,248,1);
}

.c0:hover {
  color: rgba(26,122,248,1);
}

.c0:active {
  color: rgba(77,154,255,1);
}

<span
    className="c0"
    onClick={[Function]}
  >
    Test
  </span>,
  .c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: none;
  color: rgba(102,102,102,1);
  background-color: rgba(248,248,248,1);
  -webkit-transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  transition: all 150ms cubic-bezier(0.4,0,0.2,1);
  padding: 0 10px;
  margin-right: 8px;
  font-size: 12px;
  line-height: 1;
  height: 24px;
  white-space: nowrap;
  border-radius: 999rem;
  cursor: default;
  vertical-align: middle;
  color: rgba(102,102,102,1);
  background-color: rgba(248,248,248,1);
  color: rgba(255,255,255,1);
  background-color: rgba(26,122,248,1);
}

.c0 .c1 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c0:hover {
  color: rgba(26,122,248,1);
  background-color: rgba(26,122,248,0.12);
  opacity: 1;
}

.c0 a:hover {
  color: rgba(26,122,248,1);
}

.c0:hover {
  color: rgba(26,122,248,1);
}

.c0:active {
  color: rgba(77,154,255,1);
}

.c0:hover {
  color: rgba(255,255,255,1);
  background: rgba(51,139,255,1);
}

.c0:active {
  color: rgba(255,255,255,1);
  background: rgba(12,108,235,1);
}

<span
    className="c0"
    onClick={[Function]}
  >
    Test
  </span>,
]
`;
